geiger/src/graph/Edge.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Cell from './Cell';
+import {EdgeTypes} from './EdgeTypes';
+import {computeNodeId, computeEdgeId, computeOutPortId,
+  computeInPortId, computeAudioParamPortId} from './label';
+
+/**
+ * @typedef {Object} NodesConnection
+ * @property {string} sourceNodeId
+ * @property {string} destinationNodeId
+ * @property {string} fromChannel
+ * @property {string} toChannel
+ */
+
+/**
+ * @typedef {Object} NodeParamConnection
+ * @property {string} sourceNodeId
+ * @property {string} destinationNodeId
+ * @property {string} fromChannel
+ * @property {string} destinationParamName
+ */
+
+/**
+ * @typedef {NodesConnection | NodeParamConnection} ConnectionMessage
+ */
+
+/**
+ * @typedef {Object} Source
+ * @property {string} id - The node id.
+ * @property {string} port - The port id.
+ * @typedef {Source} Target
+ */
+
+/**
+ * @typedef {Object} Point
+ * @property {number} x
+ * @property {number} y
+ */
+
+/**
+ * @class
+ * @extends Cell
+ */
+export default class Edge extends Cell {
+  /**
+   * @constructor
+   * @param {!ConnectionMessage} message
+   * @param {!EdgeTypes} type
+   */
+  constructor(message, type) {
+    super();
+
+    this.id = null;
+    this.type = type;
+
+    /** @type {!Source} */
```

Comment:
Shouldn't this be `Node` instead of `Source`?

Reply:
An `Edge` is pointing to a specific port of a `Node`. Each port has an ID. Perhaps we can use `this.source` to point to the node, then add another property to store the port ID.

Also, when creating an edge, is it good to pass a `graph` into its constructor? So that we can find the node based on the node ID from the `graph`.